### ISSUE TITLE
Interact with the environment

### ISSUE THREAD

**Issue by Ynjxsjmh** [Shiyang-Yan/Discrete-continous-PG-for-Retrieval#4]
As far as I know, reinforcement learning is a markov process

![image](https://user-images.githubusercontent.com/40143136/147864736-b5daeb61-20b5-425c-8940-0ed161d835a6.png)

that's to say, the agent needs to interact with the environment to get next state and reward.

However, in your paper I'm confused with the environment. There seems no environment just batches of training data treated as states.

https://github.com/Shiyang-Yan/Discrete-continous-PG-for-Retrieval/blob/39fd7a81f732ae043c2ea20352a0c55b72834639/train.py#L189-L205

Is there any wrong with my understanding?

**Comment by Shiyang-Yan** [comment]
Thank you for your question. The environment is defined as everything the agent is interacting with. The agent (policy) is basically the attention-generating neural model. We don't have a physical environment as in other reinforcement learning algorithms, but very much similar to reinforcement sequential learning in image captioning or machine translation.

**Comment by Ynjxsjmh** [comment]
Thanks for your reply! For model, I think you mainly define three models: `VSRN`, `EncoderImage`(method that returns model to encode image) and `EncoderText`.

![image](https://user-images.githubusercontent.com/40143136/148213439-035ce70b-7ea4-4333-8406-7699f31bf7e6.png)

`EncoderImage` and `EncoderText` are what you drawn in your Figure 2. You introduce your proposed Discrete-continuous Action Space PG to both of them.

In class `EncoderImagePrecompAttn`
https://github.com/Shiyang-Yan/Discrete-continous-PG-for-Retrieval/blob/39fd7a81f732ae043c2ea20352a0c55b72834639/model_PG.py#L574-L607

In class `EncoderText`
https://github.com/Shiyang-Yan/Discrete-continous-PG-for-Retrieval/blob/39fd7a81f732ae043c2ea20352a0c55b72834639/model_PG.py#L708-L752

That's the part you call *Discrete Action Sampling* and *Continuous Action Sampling*. I don't understand *Discrete PG Optimisation* and *Continuous PG Optimisation* very well, I guess they are implemented in class VSRN

https://github.com/Shiyang-Yan/Discrete-continous-PG-for-Retrieval/blob/39fd7a81f732ae043c2ea20352a0c55b72834639/model_PG.py#L1150-L1164

but they seem more like the loss of `EncoderImage` and `EncoderText`.

-----

I'm also confused with state. From the variable name of actions,

https://github.com/Shiyang-Yan/Discrete-continous-PG-for-Retrieval/blob/39fd7a81f732ae043c2ea20352a0c55b72834639/model_PG.py#L1112-L1113

I guess the `EncoderImage` and `EncoderText` are the agents that take state to generate action which then interacts with the environment. So `captions` and `images` are the states.

In typical RL, the environment generates states by taking agent's policy, so there is a connection between current state and next state. That's the basic of MDP.

However, if `captions` and `images` are the states, there will be no connection between current state and next state because the sequence of `captions` and `images` can be randomly generated. Is there any misunderstanding of my opinion?

**Comment by Shiyang-Yan** [comment]
As for the codes implemented, our PG algorithm's codes are built within the VSRN module. A better way is to define a new Module, we can implement this later.

**Comment by Shiyang-Yan** [comment]
Thank you for the question, please check the new codes, 
I am looking forward to more discussions.

**Comment by Ynjxsjmh** [comment]
@Shiyang-Yan  Thanks for your update, could you explain the `cosine` variable in your commit [df1ac41](https://github.com/Shiyang-Yan/Discrete-continous-PG-for-Retrieval/commit/df1ac41a8c6f15167add6a94e833bfe85976eed0)?

I guess current model_PG.py is **Our scheme (+GloVe Embedding)**.

![image](https://user-images.githubusercontent.com/40143136/152117381-b902e376-b4a9-4884-87e4-7d4cb472503a.png)

Would you mind sharing **Baseline + Discrete PG**, **Continuous PG**, **Baseline + Our PG (Our scheme)** or **Our scheme (+PG baseline)** version since I couldn't well understand what does the loop mean in `EncoderText`, `EncoderImagePrecompAttn` and want to see the implementation of your basic RL idea.

https://github.com/Shiyang-Yan/Discrete-continous-PG-for-Retrieval/blob/46dfe0cdfa4f08d32972d3f07993cbf80804f419/model_PG.py#L631
